### ISSUE TITLE
Permission issues using Red Hat Developer Studio

### ISSUE THREAD

**Issue by mickaelistria** [redhat-developer/s2i-dotnetcore#113]
As Red Hat Developer Studio enables easy continuous deploy with `rsync`, it would be great to include `rsync` in the container to allow continuous deployment from this IDE.

**Comment by mickaelistria** [comment]
Here is the error I see in Red Hat Developer Studio trying to deploy to the dotnet-2.0 image
```
Error while publishing server dotnet-demo-2 (Service) at OpenShift 3 (console.starter-us-east-2.openshift.com).  Could not sync all pods to folder /home/mistria/devstudio-11.0.0.GA/workspace/.metadata/.plugins/org.jboss.ide.eclipse.as.core/dotnet-demo@dotnet-demo-2/deploy
Syncing com.openshift.restclient.capability.resources.IRSyncable$PodPeer@d744125 to com.openshift.restclient.capability.resources.IRSyncable$LocalPeer@76ba4fa4 failed: WARNING: cannot use rsync: rsync not available in container
ERROR: Error extracting file "./.git/objects/08/97bec7f077256f459646731bca3319b66e58ed": open /home/mistria/devstudio-11.0.0.GA/workspace/.metadata/.plugins/org.jboss.ide.eclipse.as.core/dotnet-demo@dotnet-demo-2/deploy/.git/objects/08/97bec7f077256f459646731bca3319b66e58ed: permission denied
ERROR: Error extracting tar stream
Ignoring the following flags because they only apply to rsync: --exclude
error: error extracting tar at destination directory: open /home/mistria/devstudio-11.0.0.GA/workspace/.metadata/.plugins/org.jboss.ide.eclipse.as.core/dotnet-demo@dotnet-demo-2/deploy/.git/objects/08/97bec7f077256f459646731bca3319b66e58ed: permission denied
```
@jeffmaury @bdshadow Do you think the issue is in the image or in RHDS?

**Comment by mickaelistria** [comment]
I found the issue I got was caused by the project being at the root of the Git repository. I reorganized stuff, and the issue I see now is
```
Could not sync /home/mistria/devstudio-11.0.0.GA/workspace/.metadata/.plugins/org.jboss.ide.eclipse.as.core/dotnet-demo@demo-dotnet/deploy to all pods running the service demo-dotnet
Syncing com.openshift.restclient.capability.resources.IRSyncable$LocalPeer@4fe119 to com.openshift.restclient.capability.resources.IRSyncable$PodPeer@a87f365 failed: WARNING: cannot use rsync: rsync not available in container
Ignoring the following flags because they only apply to rsync: --exclude
tar: .settings: Cannot utime: Operation not permitted
tar: .settings: Cannot change mode to rwx------: Operation not permitted
tar: Pages: Cannot utime: Operation not permitted
tar: Pages: Cannot change mode to rwx------: Operation not permitted
tar: bin/Debug/netcoreapp2.0: Cannot utime: Operation not permitted
tar: bin/Debug/netcoreapp2.0: Cannot change mode to rwx------: Operation not permitted
tar: bin/Debug: Cannot utime: Operation not permitted
tar: bin/Debug: Cannot change mode to rwx------: Operation not permitted
tar: bin/Release/netcoreapp2.0: Cannot utime: Operation not permitted
tar: bin/Release/netcoreapp2.0: Cannot change mode to rwx------: Operation not permitted
tar: bin/Release: Cannot utime: Operation not permitted
tar: bin/Release: Cannot change mode to rwx------: Operation not permitted
tar: bin: Cannot utime: Operation not permitted
tar: bin: Cannot change mode to rwx------: Operation not permitted
tar: obj/Debug/netcoreapp2.0: Cannot utime: Operation not permitted
tar: obj/Debug/netcoreapp2.0: Cannot change mode to rwx------: Operation not permitted
tar: obj/Debug: Cannot utime: Operation not permitted
tar: obj/Debug: Cannot change mode to rwx------: Operation not permitted
tar: obj/Release/netcoreapp2.0: Cannot utime: Operation not permitted
tar: obj/Release/netcoreapp2.0: Cannot change mode to rwx------: Operation not permitted
tar: obj/Release: Cannot utime: Operation not permitted
tar: obj/Release: Cannot change mode to rwx------: Operation not permitted
tar: obj: Cannot utime: Operation not permitted
tar: obj: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/css: Cannot utime: Operation not permitted
tar: wwwroot/css: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/images: Cannot utime: Operation not permitted
tar: wwwroot/images: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/js: Cannot utime: Operation not permitted
tar: wwwroot/js: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/bootstrap/dist/css: Cannot utime: Operation not permitted
tar: wwwroot/lib/bootstrap/dist/css: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/bootstrap/dist/fonts: Cannot utime: Operation not permitted
tar: wwwroot/lib/bootstrap/dist/fonts: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/bootstrap/dist/js: Cannot utime: Operation not permitted
tar: wwwroot/lib/bootstrap/dist/js: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/bootstrap/dist: Cannot utime: Operation not permitted
tar: wwwroot/lib/bootstrap/dist: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/bootstrap: Cannot utime: Operation not permitted
tar: wwwroot/lib/bootstrap: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/jquery/dist: Cannot utime: Operation not permitted
tar: wwwroot/lib/jquery/dist: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/jquery: Cannot utime: Operation not permitted
tar: wwwroot/lib/jquery: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/jquery-validation/dist: Cannot utime: Operation not permitted
tar: wwwroot/lib/jquery-validation/dist: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/jquery-validation: Cannot utime: Operation not permitted
tar: wwwroot/lib/jquery-validation: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib/jquery-validation-unobtrusive: Cannot utime: Operation not permitted
tar: wwwroot/lib/jquery-validation-unobtrusive: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot/lib: Cannot utime: Operation not permitted
tar: wwwroot/lib: Cannot change mode to rwx------: Operation not permitted
tar: wwwroot: Cannot utime: Operation not permitted
tar: wwwroot: Cannot change mode to rwx------: Operation not permitted
tar: Exiting with failure status due to previous errors
error: error extracting tar at destination directory: command terminated with exit code 2
```

**Comment by jerboaa** [comment]
@mickaelistria This looks like a permission issue. It could either be a problem in our containers or in dev studio if users don't get appropriately mapped? Hard to tell. We'd need more info:

1. Which version of the containers are you using? Please provide the image name(s).
2. How would we be able to reproduce outside of dev-studio?

**Comment by jerboaa** [comment]
FWIW, the 2.0 build image calls fix-permissions which should set perms correctly:
https://github.com/redhat-developer/s2i-dotnetcore/blob/master/2.0/build/s2i/bin/assemble#L97

**Comment by jeffmaury** [comment]
According to https://docs.openshift.org/latest/dev_guide/copy_files_to_container.html#copying-requirements the main diff between rsync and tar is that directories are not created with tar where they are with rsync. So I would recommand adding rsync to the image.

**Comment by tmds** [comment]
Is this copying into /opt/app-root/src? It looks like the w permission is missing there for group and other.
```
$ docker run -ti dotnet/dotnet-20-rhel7 ls -lah /opt/app-root
total 0
drwxrwxrwx. 6 default root 51 Aug 28 08:56 .
drwxr-xr-x. 4 root    root 32 Aug 28 08:53 ..
drwxrwxrw-. 3 default root 19 Aug 28 08:53 .pki
drwxrwxrwx. 2 default root 28 Aug 28 08:54 app
drwxrwxrwx. 2 default root 55 Aug 16 08:39 etc
drwxr-xr-x. 2 default root  6 Aug 28 08:56 src
```
https://github.com/redhat-developer/s2i-dotnetcore/pull/114 should fix this.

**Comment by jerboaa** [comment]
@mickaelistria Are there specific steps how we could reproduce this outside Devstudio?

**Comment by mickaelistria** [comment]
At the moment, I neither have the time nor deep enough knowledge in
OpenShift to provide a good answer ;)

**Comment by tmds** [comment]
This shows how usage of /opt/app-root/src can give permission errors:
```
[tmds@rhex ~]$ docker run dotnet/dotnet-20-rhel7 touch /opt/app-root/src/file
[tmds@rhex ~]$ docker run --user 10000 dotnet/dotnet-20-rhel7 touch /opt/app-root/src/file
touch: cannot touch '/opt/app-root/src/file': Permission denied
[tmds@rhex ~]$ docker run --user 10000 dotnet/dotnet-20-rhel7 touch /opt/app-root/file
```

**Comment by jerboaa** [comment]
@tmds interesting. So `src` has too restrictive permissions. Question is whether @mickaelistria's problem results from it trying to create directories/files in `src` rather than `/opt/app-root`

**Comment by mickaelistria** [comment]
From Red Hat Developers Studio, I do not have control about what ends where. It just happens to work for Java or other plain HTML projects; I don't know whether those are using the same conventions (using `src` or not)

**Comment by tmds** [comment]
@mickaelistria if it is possible, please share the (failing) build log when using the dotnet-runtime:2.0 image instead of the dotnet:2.0 image. The runtime image has a working dir with proper permissions. So oc rsync should succeed. Then the build will fail since there is no sdk in that image. This would confirm the rsync fail is due to the permissions on the /opt/app-root/src folder.

**Comment by mickaelistria** [comment]
I can't find a way to use the `dotnet-runtime:2.0` image instead of the
`dotnet:2.0` image, neither from Red Hat Developers Studio nor from Web
Console. I'm afraid I'm not good enough at openshift stuff to give you more
help.
Looking at the various logs, I have the impression that the permissiong on
/opt/app-root/src folders are indeed the cause.

**Comment by mickaelistria** [comment]
How can I locally change the https://github.com/redhat-developer/s2i-dotnetcore/blob/master/2.0/build/Dockerfile.rhel7 to update permission and then deploy it in my OpenShift project so I can use it for my application?

**Comment by tmds** [comment]
> How can I locally change the https://github.com/redhat-developer/s2i-dotnetcore/blob/master/2.0/build/Dockerfile.rhel7 to update permission

Change "mkdir /opt/app-root/src" to "mkdir /opt/app-root/src && fix-permissions /opt/app-root/src".

> and then deploy it in my OpenShift project so I can use it for my application?

I'd have to figure it out myself. @jerboaa ?

**Comment by jerboaa** [comment]
@mickaelistria try forking https://github.com/jerboaa/s2i-dotnetcore, make your changes to the Dockerfile and then use:

     oc new-build --context-dir=2.0/build https://github.com/mickaelistria/s2i-dotnetcore

Might want to opt for adding `--to=dotnet-mickael` or some such. Then use `dotnet-mickael` for your test. This assumes the host is properly subscribed to relevant channels via subscription-manager.

**Comment by aslicerh** [comment]
The permissions part of this problem should be fixed by PR #114.

@mickaelistria: Please let us know if you're still seeing issues, either by (re)opening an issue or reaching out on IRC.

**Comment by mickaelistria** [comment]
I'm still seeing issues, but those seems to be more on the DevStudio/Git side. I believe #118 would help.
